File: lab2/max.cpp
#define CATCH_CONFIG_MAIN
#include "catch.hpp"
#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

int find_max(const vector<int> &arr, int low, int high) {
  if (low == high) {
    return arr[low];
  }

  int mid = (low + high) / 2;
  int max_left = find_max(arr, low, mid);
  int max_right = find_max(arr, mid + 1, high);

  return max(max_left, max_right);
}

vector<int> read_data(const string &filename) {
  ifstream file(filename);
  int n;
  file >> n;

  vector<int> arr(n);
  for (int i = 0; i < n; ++i) {
    file >> arr[i];
  }
  return arr;
}

TEST_CASE("test", "[find_max]") {
  SECTION("Test1") {
    vector<int> arr = read_data("../max1.txt");
    int result = find_max(arr, 0, arr.size() - 1);
    REQUIRE(result == 5);
  }

  SECTION("Test2") {
    vector<int> arr = read_data("../max2.txt");
    int result = find_max(arr, 0, arr.size() - 1);
    REQUIRE(result == -1);
  }
}

File: lab2/max.py
import unittest


def find_max(arr, low, high):
    if low == high:
        return arr[low]

    mid = (low + high) // 2

    max_left = find_max(arr, low, mid)
    max_right = find_max(arr, mid + 1, high)

    return max(max_left, max_right)


def read_data(filename):
    with open(filename, "r") as file:
        line = file.readline().strip()
        array_part, expected_part = line.split("),(")
        arr = list(map(int, array_part.strip("()").split(",")))
        expected = int(expected_part.strip("()"))
    return arr, expected


class Test(unittest.TestCase):

    def test1(self):
        arr, expected = read_data("../max1.txt")
        result = find_max(arr, 0, len(arr) - 1)
        self.assertEqual(result, expected)

    def test2(self):
        arr, expected = read_data("../max2.txt")
        result = find_max(arr, 0, len(arr) - 1)
        self.assertEqual(result, expected)


if __name__ == "__main__":
    unittest.main()
